File: mcp-code-intelligence-python/src/mcp_code_intel/analyzers/graph_analysis/circular_dep_detector.py
"""KSA-163: Circular Dependency Detector."""
from __future__ import annotations
from . import CircularDep, CycleChain, CycleNode, AdjacencyList
from .utils import GraphLoader, TarjanSCC


class CircularDepDetector:
    def __init__(self, graph_loader: GraphLoader) -> None:
        self._gl = graph_loader

    def detect(self, module: str | None = None, max_length: int | None = None) -> list[CircularDep]:
        graph = self._gl.load_dependency_graph(module)
        if not graph:
            return []
        sccs = TarjanSCC().find_sccs(graph)
        results = []
        for scc in sccs:
            if max_length and len(scc) > max_length:
                continue
            results.append(self._build(scc, graph, module))
        sev_order = {"high": 0, "medium": 1, "low": 2}
        return sorted(results, key=lambda d: (sev_order.get(d.severity, 3), d.length))

    def _build(self, scc: list[int], graph: AdjacencyList, module: str | None) -> CircularDep:
        infos = self._gl.get_symbol_info_batch(scc)
        scc_set = set(scc)
        ordered = self._order_cycle(scc, graph, scc_set)
        nodes = [CycleNode(id, infos.get(id, None) and infos[id].name or f"sym_{id}",
                           infos.get(id, None) and infos[id].file_path or "unknown",
                           infos.get(id, None) and infos[id].kind or "unknown") for id in ordered]
        edges = [f"{infos.get(ordered[i], None) and infos[ordered[i]].name or ordered[i]} → "
                 f"{infos.get(ordered[(i+1)%len(ordered)], None) and infos[ordered[(i+1)%len(ordered)]].name or ordered[(i+1)%len(ordered)]}"
                 for i in range(len(ordered))]
        severity = "high" if len(scc) <= 2 else "medium" if len(scc) <= 4 else "low"
        return CircularDep(CycleChain(nodes, edges), len(scc), severity, module)

    def _order_cycle(self, scc: list[int], graph: AdjacencyList, scc_set: set[int]) -> list[int]:
        visited: set[int] = set()
        ordered: list[int] = []
        current = scc[0]
        while current not in visited:
            visited.add(current); ordered.append(current)
            nxt = next((n for n in graph.get(current, []) if n in scc_set and n not in visited), None)
            if nxt is None: break
            current = nxt
        return ordered if len(ordered) == len(scc) else scc

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/graphanalysis/CircularDepDetector.kt
/**
 * KSA-163: Circular Dependency Detector using Tarjan's SCC.
 */
package com.codeintel.analyzers.graphanalysis

import com.codeintel.analyzers.graphanalysis.utils.GraphLoader
import com.codeintel.analyzers.graphanalysis.utils.TarjanSCC

class CircularDepDetector(private val graphLoader: GraphLoader) {

    fun detect(options: CircularDepOptions = CircularDepOptions()): List<CircularDep> {
        val graph = graphLoader.loadDependencyGraph(options.module)
        if (graph.isEmpty()) return emptyList()

        val tarjan = TarjanSCC()
        val sccs = tarjan.findSCCs(graph)

        return sccs
            .filter { options.maxLength == null || it.size <= options.maxLength }
            .map { scc -> buildCircularDep(scc, graph, options.module) }
            .sortedWith(compareBy({ severityOrder(it.severity) }, { it.length }))
    }

    private fun buildCircularDep(scc: List<Int>, graph: AdjacencyList, module: String?): CircularDep {
        val infos = graphLoader.getSymbolInfoBatch(scc)
        val sccSet = scc.toSet()
        val ordered = orderCycle(scc, graph, sccSet)
        val nodes = ordered.map { id ->
            val info = infos[id]
            CycleNode(id, info?.name ?: "symbol_$id", info?.filePath ?: "unknown", info?.kind ?: "unknown")
        }
        val edges = ordered.indices.map { i ->
            val curr = infos[ordered[i]]?.name ?: "${ordered[i]}"
            val next = infos[ordered[(i + 1) % ordered.size]]?.name ?: "${ordered[(i + 1) % ordered.size]}"
            "$curr → $next"
        }
        return CircularDep(CycleChain(nodes, edges), scc.size, classifySeverity(scc.size), module)
    }

    private fun orderCycle(scc: List<Int>, graph: AdjacencyList, sccSet: Set<Int>): List<Int> {
        val visited = mutableSetOf<Int>()
        val ordered = mutableListOf<Int>()
        var current = scc[0]
        while (current !in visited) {
            visited.add(current); ordered.add(current)
            val next = graph[current]?.firstOrNull { it in sccSet && it !in visited }
            current = next ?: break
        }
        return if (ordered.size == scc.size) ordered else scc
    }

    private fun classifySeverity(length: Int): String = when {
        length <= 2 -> "high"; length <= 4 -> "medium"; else -> "low"
    }

    private fun severityOrder(s: String): Int = when (s) {
        "high" -> 0; "medium" -> 1; else -> 2
    }
}

data class CircularDepOptions(val module: String? = null, val maxLength: Int? = null)
